### PR TITLE
Add driver is31fl3194

### PR DESCRIPTION
**Pull request by benjaminbjornsson**

### CODE REVIEW

**Comment by carlescufi** [comment]
@mbolivar-nordic @MaureenHelm could you take a look?

File: dts/bindings/led/issi,is31fl3194.yaml
Comment (simonguinot) [inline dts/bindings/led/issi,is31fl3194.yaml:11]: I don't think it is the right description :)

**Comment by simonguinot** [comment]
Hi @carlescufi. I'll review it in the next few days.

**Comment by benjaminbjornsson** [comment]
> Hi @carlescufi. I'll review it in the next few days.

Thank you @simonguinot !

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic [...] could you take a look?

I will try, but this is led, not led_strip, so @Mani-Sadhasivam is the maintainer

**Comment by Mani-Sadhasivam** [comment]
It's on the pipeline for review this weekend.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [inline drivers/led/is31fl3194/is31fl3194.c:27]: I am not sure, but I think this could be `const` because the pointers inside are not meant to be written. Isn't it ?

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [inline drivers/led/is31fl3194/is31fl3194.c:146]: Should you not implement the set_color function ? It looks like configuring the color is supported by the IS31FL3194 LED driver.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [inline drivers/led/is31fl3194/is31fl3194.c:106]: In led.h it is written that the brightness value is a percent. So I think you have to convert the brightness value from the [0-100] range to [0-255].

Note that the API should be changed because there is no reason to loose granularity on brightness.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [inline drivers/led/is31fl3194/is31fl3194.c:208]: I don't think it is working. The property is named `curr-band-max` in the DT binding. So it should be `curr_band_max` here.

File: dts/bindings/led/issi,is31fl3194.yaml
Comment (simonguinot) [inline dts/bindings/led/issi,is31fl3194.yaml:24]: Please, can you add some description for this property ?

File: drivers/led/is31fl3194/is31fl3194_reg.h
Comment (simonguinot) [inline drivers/led/is31fl3194/is31fl3194_reg.h:97]: I think you can remove the `: 8` here :)

File: drivers/led/is31fl3194/is31fl3194_reg.c
Comment (simonguinot) [inline drivers/led/is31fl3194/is31fl3194_reg.c:116]: As I said in another comment, I don't see the purpose for this indirection.

**Review by simonguinot — DISMISSED** [review]
Hi @benjaminbjornsson,

You really did a nice job with this driver. The IS31FL3194 is a beast. I can't wait to see how you will implement the hardware LED effects (such as breathing). Probably you already have some ideas how we could add support for LED effects to the LED API.

Note that I didn't review completely this PR. It was too big for the time I had. If needed I'll do another pass.

See below for my comments.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (benjaminbjornsson) [reply drivers/led/is31fl3194/is31fl3194.c:56]: Agree, it currently just enables it. Does turning it on mean that it should be set to full brightness?

File: drivers/led/is31fl3194/is31fl3194.c
Comment (benjaminbjornsson) [reply drivers/led/is31fl3194/is31fl3194.c:106]: Agreed, will updated this.

I agree, makes more sense to use full range of `uint8_t`.

File: dts/bindings/led/issi,is31fl3194.yaml
Comment (benjaminbjornsson) [reply dts/bindings/led/issi,is31fl3194.yaml:11]: Yeah, that doen't look right... Thank you!

File: drivers/led/is31fl3194/is31fl3194.c
Comment (benjaminbjornsson) [reply drivers/led/is31fl3194/is31fl3194.c:53]: Agreed, will add const here.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (benjaminbjornsson) [reply drivers/led/is31fl3194/is31fl3194.c:27]: I don't think const is necessary inside `struct is31fl3194_config`. The entire struct will be declared as `const` when used in the device macro.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (benjaminbjornsson) [reply drivers/led/is31fl3194/is31fl3194.c:146]: I think the reason why I didn't is that `set_color` sets the color of a certain LED whereas this driver has three single color LEDs connected to it.

File: drivers/led/is31fl3194/is31fl3194_reg.h
Comment (benjaminbjornsson) [reply drivers/led/is31fl3194/is31fl3194_reg.h:97]: I think it's quite nice to state explicitly that we're using the whole `uint8_t` here and IMO it makes sense to be consistent across structs.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [reply drivers/led/is31fl3194/is31fl3194.c:56]: Since it is not specified, it is up to you. But I think that full brightness should be right. Note that I am not a big fan of the `led_off` / `led_on` functions. They don't exist in the Linux LED API for example. I think of them pretty much as wrappers to `led_set_brightness` with brightness set to 0 or 100.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [reply drivers/led/is31fl3194/is31fl3194.c:27]: OK

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [reply drivers/led/is31fl3194/is31fl3194.c:146]: Are there 3 LEDs ? Or a single LED with 3 channels ?

In the former case, a user may need to configure the LED by calling a single function (with a single I2C command) rather than setting the brightness of three different LED which are in fact a single physical one. This makes a difference when running software LED effects.

Note that the write_channels function can also be used to do that, but in a much less intuitive way.

**Comment by simonguinot** [comment]
@benjaminbjornsson I think it would be nice to provides a DTS overlay example along with the sample (maybe for the board you are using to test the driver).

File: drivers/led/is31fl3194/is31fl3194_reg.h
Comment (simonguinot) [reply drivers/led/is31fl3194/is31fl3194_reg.h:97]: OK

**Comment by benjaminbjornsson** [comment]
I've updated the PR and I think I've fixed all issues raised by @simonguinot. Decided to remove the auto-breathing functionality from the abstraction layer to make the PR easier to review.

**Comment by benjaminbjornsson** [comment]
> @benjaminbjornsson I think it would be nice to provides a DTS overlay example along with the sample (maybe for the board you are using to test the driver).

Isn't the node provided in the Arduino Nicla Sense ME board DTS that is part of this PR enough as an example? Or should that node be moved into an overlay of the example?

**Comment by simonguinot** [comment]
> > @benjaminbjornsson I think it would be nice to provides a DTS overlay example along with the sample (maybe for the board you are using to test the driver).
> 
> Isn't the node provided in the Arduino Nicla Sense ME board DTS that is part of this PR enough as an example? Or should that node be moved into an overlay of the example?

Yes it is enough.

File: drivers/led/is31fl3194/is31fl3194.c
Comment (simonguinot) [reply drivers/led/is31fl3194/is31fl3194.c:134]: According to the datasheet the is31fl3194 is a 3-channel LED driver and both options are possible: 3 single color LEDs, or a RGB / RG+W LED. So maybe you should implement `set_color` as well. It is your call and you can consider this comment resolved either way.

**Comment by simonguinot** [comment]
Sorry, I closed this PR by accident.

**Review by simonguinot — COMMENTED** [review]
@benjaminbjornsson I still don't like the HAL idea and especially the indirection around the I2C functions (i.e. `ctx->{read,write}_reg()` -> `is31fl3194_i2c_{read,write}()` -> `i2c_burst_{read,write}_dt()`). IMHO it only makes sense if you plan to use your HAL with another OS. But I can't see the gain for Zephyr. This complicates the code. And my advice is still that you should simplify the driver.

I'll be in vacations for the next three weeks and I don't want to block this PR. So I am withdrawing my initial "request changes" feedback. But I am not approving it either.

**Comment by simonguinot** [comment]
Hi @Mani-Sadhasivam. Please can you have a look ? What do you think of the "I2C abstraction" code in `is31fl3194_reg.{c,h}` ?

File: drivers/led/is31fl3194/is31fl3194_reg.c
Comment (Mani-Sadhasivam) [reply drivers/led/is31fl3194/is31fl3194_reg.c:33]: NACK. These kind of HAL abstractions makes sense if you are creating a library that might be used by different drivers. But here, there is obviously one driver making use of it and the HAL indirection is really making the code look complex. Please get rid of it.

**Review by Mani-Sadhasivam — CHANGES_REQUESTED** [review]

**Comment by Mani-Sadhasivam** [comment]
> Hi @Mani-Sadhasivam. Please can you have a look ? What do you think of the "I2C abstraction" code in `is31fl3194_reg.{c,h}` ?

Thanks, @simonguinot for the reviews, really appreciate your help :)  As I said in the review, I'm not in favour of adding the HAL abstraction to this driver. So that part needs to be reworked.

**Pull request closed without merging**
